egs/wsj/s5/utils/lang/make_subword_lexicon_fst.py

```diff
@@ -0,0 +1,159 @@
+        form 'word prob p1 p2...', e.g. 'a, 1.0 ay'""")
+    args = parser.parse_args()
+    return args
+
+def write_fst_no_silence(lexicon):
+    pass
+
+def contain_disambig_symbol(phones):
+    return True if phones[-1][0] == "#" else False
+
+
+def write_fst_with_silence(lexicon, sil_phone, sil_prob, sil_disambig, position_dependent, separator):
+
+    def print_arc(src, dest, phone, word, cost):
+        print('{}\t{}\t{}\t{}\t{}'.format(src, dest, phone, word, cost))
+
+    def is_end(word, separator):
+        return (not word.endswith(separator))
+
+    def get_suffix(phone):
+        return phone[-2:]
+
+    def get_position(phones):
+        phones = phones[:-1] if contain_disambig_symbol(phones) else phones
+
+        if len(phones) == 1:
+            suffix = get_suffix(phones[0])
+            if suffix == "_B":
+                return "begin"
+            elif suffix == "_I":
+                return "middle"
+            elif suffix == "_E":
+                return "end"
+            else:
+                return "word"
+        else:
+            if get_suffix(phones[0]) == "_B":
+                if get_suffix(phones[-1]) == "_I":
+                    return "begin"
+                elif get_suffix(phones[-1]) == "_E":
+                    return "word"
+            elif get_suffix(phones[0]) == "_I":
+                if get_suffix(phones[-1]) == "_I":
+                    return "middle"
+                elif get_suffix(phones[-1]) == "_E":
+                    return "end"
+
+    def print_word(word, pron_cost, phones, position, next_state):
+    
+        assert(len(phones)) >= 0
+    
+        from_state = prefix_state
+        to_state = prefix_state
+        if position == "begin":
+            from_state = loop_state
+        elif position == "end": 
+            to_state = "sil_nonsil"
+        elif position == "word":
+            from_state = loop_state
+            to_state = "sil_nonsil"
+    
+        current_state = from_state
+        for i in range(len(phones) - 1):
+            word = (word if i == 0 else '<eps>')
+            cost = (pron_cost if i == 0 else 0.0)
+            print_arc(current_state, next_state, phones[i], word, cost)
+            current_state = next_state
+            next_state += 1
+    
+        i = len(phones) - 1
+        phone = (phones[i] if i >= 0 else "<eps>")
+        word = (word if i <= 0 else "<eps>")
+        cost = (pron_cost if i <= 0 else 0.0)
+        if to_state != "sil_nonsil":
+            print_arc(current_state, to_state, phone, word, cost)
+        else:
+            p_cost = cost
+            cost = no_sil_cost + p_cost
+            print_arc(current_state, loop_state, phone, word, cost)
+            cost = sil_cost + p_cost
+            print_arc(current_state, sil_state, phone, word, cost)
+    
+        return next_state
+
+    sil_cost = -math.log(sil_prob)
+    no_sil_cost = -math.log(1 - sil_prob)
```

Comment:
what if sil_prob = 1.0? you should either check that before calling this function and print some warning; or assign an "infinity" value here